docs/proposals/ldap-group-sync.md

```diff
@@ -0,0 +1,118 @@
+
+```go
+// LDAPGroupDataExtractor retrieves data about an LDAP group from the LDAP server.
+type LDAPGroupDataExtractor interface {
+	// ExtractMembers returns the list of LDAP first-class user entries that are members of the LDAP group
+	// specified by the groupUID
+	ExtractMembers(groupUID string) (members []*ldap.Entry, err error)
+	// ExtractAnnotations returns the mapping of annotation keys to values that the OpenShift group Representing
+	// the LDAP group specified by the groupUID should have
+	ExtractAnnotations(groupUID string) (annotations map[string]string, err error)
+
+	LDAPGroupNameMapper
+}
+
+// LDAPGroupNameMapper maps a groupUID representing an LDAP group to the OpenShift Group name for the resource
+type LDAPGroupNameMapper interface {
+	GroupNameFor(groupUID string) (groupName string, err error)
+}
+```
+
+## Determining OpenShift `Identities` for LDAP Members
+This will be done in an identical way to the mapping for [LDAP password processing](https://github.com/openshift/origin/blob/master/pkg/auth/authenticator/password/ldappassword/ldap.go#L227-243) once the requisite data is collected.
+
+```go
+// LDAPIdentityFactory creates Identites for LDAP user entries.
+type LDAPUserIdentityFactory interface {
+	IdentityFor(user *ldap.Entry) (identity userapi.Identity, err error)
+}
+```
+
+## Determining OpenShift `Users` for OpenShift `Identities`
+A new `UserIdentityMapper` will need to be created: the `DeterministicUserIdentityMapper`, which will assume that the unique identifier retrieved from the LDAP response and integrated into the OpenShift `Identity` will be deterministicically mapped to an OpenShift `User`. This is required as otherwise there is no deterministic mapping from the `Identity` created and any OpenShift `User`.
+
+```go
+// DeterministicUserIdentityMapper is a UserIdentityMapper that forces a deterministic mapping from
+// an Identity to a User.
+type DeterministicUserIdentityMapper struct {
+  // AllowIdentityCollisions determines if this UserIdentityMapper is allowed to make a mapping
+  // between an Identity provided by LDAP and a User that already is mapped to another Identity.
+  AllowIdentityCollisions bool
+}
+
+func (m *DeterministicUserIdentityMapper) UserFor(identity authapi.UserIdentityInfo) (user user.Info, err error) {
+}
+```
+
+## Populating OpenShift `Group` Records With Data From Sync Job
+Given the LDAP group UID, the corresponding OpenShift `Group` can either be found with the OpenShift `Group` name from the `LDAPGroupMapper` or created it it does not yet exist. If the group returned from the LDAP server is a first-class entry, information can be migrated to the OpenShift `Group`, but if the group is encode as attributes of first-class LDAP user entries, no information other than the name of the group can be carried over. In either case, data from the old OpenShift `Group`'s `ObjectMeta` is kept unless overridden by the update and the `Group`'s `Users` are always overriden.
```

Comment:
1. get existing openshift group or start a new one
1. set annotations and labels
1. set openshift users
1. replace or create openshift group

I think that covers everything here.